File: ports/kotlin/src/commonMain/kotlin/com/codinglombok/lombokalgoritma/ML.kt
// LombokAlgoritma — Kotlin ML Module
// Apache-2.0 — @codinglombok
package com.codinglombok.lombokalgoritma
import kotlin.math.sqrt

fun dotProduct(a: DoubleArray, b: DoubleArray): Double = a.zip(b.asList()).sumOf { (x,y) -> x*y }
fun l2Norm(v: DoubleArray): Double = sqrt(v.sumOf { it*it })
fun cosineSimilarity(a: DoubleArray, b: DoubleArray): Double {
    val na = l2Norm(a); val nb = l2Norm(b)
    return if (na == 0.0 || nb == 0.0) 0.0 else dotProduct(a, b) / (na * nb)
}
fun l2Distance(a: DoubleArray, b: DoubleArray): Double = sqrt(a.zip(b.asList()).sumOf { (x,y) -> (x-y)*(x-y) })
fun normalize(v: DoubleArray): DoubleArray { val n=l2Norm(v); return if(n==0.0)DoubleArray(v.size) else DoubleArray(v.size){v[it]/n} }

File: ports/swift/Sources/LombokAlgoritma/ML.swift
// LombokAlgoritma — Swift ML/Vector Module
// Apache-2.0 — @codinglombok
import Foundation

public enum VectorOps {

    public static func dotProduct(_ a: [Double], _ b: [Double]) -> Double {
        zip(a, b).reduce(0) { $0 + $1.0 * $1.1 }
    }

    public static func l2Norm(_ v: [Double]) -> Double {
        sqrt(v.reduce(0) { $0 + $1 * $1 })
    }

    public static func cosineSimilarity(_ a: [Double], _ b: [Double]) -> Double {
        let na = l2Norm(a), nb = l2Norm(b)
        guard na > 0 && nb > 0 else { return 0 }
        return dotProduct(a, b) / (na * nb)
    }

    public static func l2Distance(_ a: [Double], _ b: [Double]) -> Double {
        sqrt(zip(a, b).reduce(0) { $0 + pow($1.0 - $1.1, 2) })
    }

    public static func normalize(_ v: [Double]) -> [Double] {
        let n = l2Norm(v)
        guard n > 0 else { return [Double](repeating: 0, count: v.count) }
        return v.map { $0 / n }
    }
}
